src/coreclr/jit/morph.cpp

```diff
@@ -12789,6 +12794,11 @@ GenTree* Compiler::fgMorphModToZero(GenTreeOp* tree)
     gtExtractSideEffList(op1, &op1SideEffects, GTF_ALL_EFFECT);
     if (op1SideEffects != nullptr)
     {
+        // Do not transform this if there are side effects and we are not in global morph.
+        // If we want to allow this, we need to update value numbers for the GT_COMMA.
```

Comment:
`gtExtractSideEffList` alters the state of IR, so this check has to be done before using it.

Reply:
That's unfortunate. I'll add the check before then.

Reply:
If I add the check before, then the optimization when there are no side effects will not happen when not in global morph. So this makes this a little more complicated.

Reply:
Why is `gtExtractSideEffList` actually causing a side effect itself? This should just be a function to get information, not modify it.

Reply:
(I would expect you could check it like this: `if ((gtFlags & GTF_SIDE_EFFECT) == 0) { /* no side effects */ } else if (fgGlobalMorph) { /* side effects and in global morph */ }`.)

Reply:
> Why is `gtExtractSideEffList` actually causing a side effect itself? This should just be a function to get information, not modify it.

Just a consequence of the how it is currently used. It is unusual to have the situation we have here, plus, the side effects flags can always be checked instead to get approximately the same answer.

Reply:
I'm going to have to think on this. The code gets messy when I want to use `gtExtractSideEffList`. Might be better off just checking the flag and wrapping `op1` with a new tmp and comma.

Reply:
I am not sure what the point of temporaries would be. In my mind the code could look like this:
```scala
if (((op1->gtFlags & GTF_SIDE_EFFECT) != 0) && !fgGlobalMorph)
{
    // Comment about VNs.
    retrun nullptr;
}

op2->AsIntConCommon()->SetIconValue(0);
fgUpdateConstTreeValueNumber(op2);

GenTree* op1SideEffects = nullptr;
gtExtractSideEffList(op1, &op1SideEffects, GTF_ALL_EFFECT);
if (op1SideEffects != nullptr) // Same code as we already have below
```

Reply:
I see, that looks fine since we will assume extracting side effects will always be null if that flag is not flag. I was overthinking it.